content/blog/2016-07-23-pull-does-not-scale-or-does-it.md

```diff
@@ -0,0 +1,178 @@
+system immediately as it happens. This central system then either aggregates
+the events into metrics (StatsD is one example of this) or stores events
+individually for later processing (the ELK stack is an example of that). In
+such a system, pulling would be problematic indeed: the instrumented service
+would have to buffer events between pulls, and the pulls would have to happen
+incredibly frequently in order to simulate the same “liveness” of the
+push-based approach and not overwhelm event buffers.
+
+However, again, Prometheus is not an event-based monitoring system. You do not
+send raw events to Prometheus, nor can it store them. Prometheus is in the
+business of collecting aggregated time series data. That means that it's only
+interested in regularly collecting the current *state* of a given set of
+metrics, not the underlying events that led to the generation of those metrics.
+For example, an instrumented service would not send a message about each HTTP
+request to Prometheus as it is handled, but would simply count up those
+requests in memory.  This can happen hundreds of thousands of times per second
+without causing any monitoring traffic. Prometheus then simply asks the service
+instance every 15 or 30 seconds (or whatever you configure) about the current
+counter value and stores that value together with the scrape timestamp as a
+sample. Other metric types, such as gauges, histograms, and summaries, are
+handled similarly. The resulting monitoring traffic is low, and the pull-based
+approach also does not create problems in this case.
+
+## But now my monitoring needs to know about my service instances!
+
+With a pull-based approach, your monitoring system needs to know which service
+instances exist and how to connect to them. Often people are worried about the
+extra configuration this requires on the side of the monitoring system.
+
+We would argue that you cannot escape this configuration effort for
+serious monitoring setups in any case: if your monitoring system doesn't know
+what the world *should* look like and which monitored service instances
+*should* be there, how would it be able to tell when an instance just never
+reports in, is down due to an outage, or really is no longer meant to exist?
+This is only acceptable if you never care about the health of individual
+instances at all, like when you only run ephemeral workers where it is
+sufficient for a large-enough number of them to report in some result. Most
+environments are not exclusively like that.
+
+If the monitoring system needs to know the desired state of the world anyway,
+then a push-based approach actually requires *more* configuration in total. Not
+only does your monitoring system need to know what service instances should
+exist, but your service instances now also need to know how to reach your
+monitoring system. This is not only more required configuration, it also makes
+your monitoring setup less flexible. Now you cannot just run a copy of
```

Comment:
I'd frame this as a positive for pull rather than a negative for push.

I think it'd help convince people more to show the cool stuff you can do with pull (e.g. decentralisation) rather than saying what push can't do.